### ISSUE TITLE
Document server.py functions

### ISSUE THREAD

**Issue by mwest1066** [PrairieLearn/PrairieLearn#2052]
Much of this can be copied from `devElements.md`, but other functions like `file()` also need to be added.

**Comment by tdy** [comment]
Here's my current limited understanding of `file()`:

Function | Return object | modifiable `data` keys | unmodifiable `data` keys | Description
-- | -- | -- | -- | --
`file()` | Seems like it returns a memory buffer, but I'm also seeing an example where it returns a string (not a string buffer?) | `params`, `filename`, **any others?** | **?** | Generate a file object dynamically in lieu of a physical file. Trigger via `type="dynamic"` in the question element (e.g., `pl-figure`, `pl-file-download`). Access the requested filename via `data['filename']`.
**copy rest of `elements` table here?** | -- | -- | -- | --
**or just link to it?** | -- | -- | -- | --

**Comment by coatless** [comment]
Elements table:

https://prairielearn.readthedocs.io/en/latest/devElements/#element-functions

How `file()` function is used with `pl-figure` w/ examples:

https://prairielearn.readthedocs.io/en/latest/elements/#pl-figure-element

**Comment by mwest1066** [comment]
Generated files are handled by this route: https://github.com/PrairieLearn/PrairieLearn/blob/master/server.js#L629

Which then is handled by: https://github.com/PrairieLearn/PrairieLearn/blob/master/pages/studentGeneratedFilesQuestion/studentGeneratedFilesQuestion.js#L22

Which then calls `getFile()`: https://github.com/PrairieLearn/PrairieLearn/blob/master/lib/question.js#L130

Which calls `file()`: https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L1067

This then calls the functions to process the question HTML with `phase='file'`. The main processing function is `traverseQuestionAndExecuteFunctions()`: https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L384

This calls the actual element `file()` function: https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L405

The result is processed specially for `phase = 'file'`: https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L427

Here we use the NodeJS Buffer module: https://nodejs.org/api/buffer.html

This expects `ret_val` to either be a NodeJS Buffer or a base64 string.

No matter what `ret_val` is, we convert it to a Buffer and save it in `fileData`. This `fileData` variable gets passed all the way back up the call chain and finally returned to the client using `res.send()`: https://github.com/PrairieLearn/PrairieLearn/blob/master/pages/studentGeneratedFilesQuestion/studentGeneratedFilesQuestion.js#L25

To understand what type `ret_val` will be, we can dig down further into how we call the actual element Python function. This starts with `elementFunction()`: https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L166

Here `pc` is a `PythonCaller` object: https://github.com/PrairieLearn/PrairieLearn/blob/master/lib/code-caller.js#L103

So we call `pc.call()`, which after some amount of bouncing around ends up in the Python zygote code: https://github.com/PrairieLearn/PrairieLearn/blob/master/lib/python-caller-trampoline.py#L42

This calls the Python element function: https://github.com/PrairieLearn/PrairieLearn/blob/master/lib/python-caller-trampoline.py#L104

The return value of the Python function is handled specially for `file()` calls: https://github.com/PrairieLearn/PrairieLearn/blob/master/lib/python-caller-trampoline.py#L115

From this it looks like we will do our best to accept various formats, treat them as binary data, and encode them as a base64 string.

So in summary, the `file()` function can return anything handled by https://github.com/PrairieLearn/PrairieLearn/blob/master/lib/python-caller-trampoline.py#L115

This will be base64-encoded, returned to the main PL process, which will base64 decode it and return the resulting binary data to the client.

**Comment by mwest1066** [comment]
To understand what can be modified in which phases, check out https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L361
